snapcraft/internal/lifecycle/_runner.py

```diff
@@ -152,83 +159,171 @@ def __init__(self, config, project_options):
             # and keep using a list.
             parts = [p for p in self.config.all_parts if p.name in part_names]
+            processed_part_names = part_names
         else:
             parts = self.config.all_parts
-            part_names = self.config.part_names
+            processed_part_names = self.config.part_names
 
         step_index = common.COMMAND_ORDER.index(step) + 1
 
-        for step in common.COMMAND_ORDER[0:step_index]:
-            if step == 'stage':
-                # XXX check only for collisions on the parts that have already
-                # been built --elopio - 20170713
-                pluginhandler.check_for_collisions(self.config.all_parts)
-            for part in parts:
-                if step not in self._steps_run[part.name]:
-                    self._run_step(step, part, part_names)
-                    self._steps_run[part.name].add(step)
+        with config.CLIConfig() as cli_config:
+            for current_step in common.COMMAND_ORDER[0:step_index]:
+                if current_step == 'stage':
+                    # XXX check only for collisions on the parts that have
+                    # already been built --elopio - 20170713
+                    pluginhandler.check_for_collisions(self.config.all_parts)
+                for part in parts:
+                    if self._dirty_reports[part.name][current_step]:
+                        self._handle_dirty(
+                            part, current_step,
+                            self._dirty_reports[part.name][current_step],
+                            cli_config)
+                    elif current_step in self._steps_run[part.name]:
+                        # By default, if a step has already run, don't run it
+                        # again. However, automatically clean and re-run the
+                        # step if all the following conditions apply:
+                        #
+                        #   1. The step is the exact step that was requested
+                        #      (not an earlier one)
+                        #   2. The part was explicitly specified
+                        if (part_names and current_step == step and
+                                part.name in part_names):
+                            getattr(self, '_re{}'.format(current_step))(part)
+                        else:
+                            notify_part_progress(
+                                part,
+                                'Skipping {}'.format(current_step),
+                                '(already ran)')
+                    else:
+                        getattr(self, '_run_{}'.format(current_step))(part)
+                        self._steps_run[part.name].add(current_step)
+
+        self._create_meta(step, processed_part_names)
+
+    def _run_pull(self, part):
+        self._run_step(step='pull', part=part, progress='Pulling')
 
-        self._create_meta(step, part_names)
+    def _repull(self, part, hint=''):
```

Comment:
seems you stopped using `hint` completely, I might be wrong, but I haven't seen any uses of it, why not just remove it?

Reply:
It is still being used, actually.